README.md
# Yeah! just one letter ! [![Build Status](https://travis-ci.org/mssola/g.svg)](https://travis-ci.org/mssola/g)

This repository holds a very humble and simple script that handles shortcuts
of bash paths. For example:

    $ g awesome
    (instead of ...)
    $ cd /home/user/directory/dir/awesome

Furthermore, this repo also brings a file that adds bash completion so the user
can access the desired directory as fast as possible. In order to manage
shortcuts, this script accepts three commands:

    $ g add awesome /home/user/directory/dir/awesome
    $ g rm awesome
    $ g list

I guess that the commands above are quite self-explanatory. Note though that if
the path was not given to the add command then the current working directory
will be picked instead.

## Install

Just place the `g.sh` and the `gcompletion.sh` files wherever you want and then
add a `source` command for both of the files in the `.bashrc` file.

## Legal stuff

```
Copyright (C) 2013-2020 Miquel Sabaté Solà <mikisabate@gmail.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
```


CHANGELOG.org
#+STARTUP:showall

* Changelog

** 0.3.x

Lots of minor code improvements across different releases that were not tagged
in git:

- a54d534d3585 Improved the layout of the table displayed by the =list= command.
- 3b18f5039369 Improved the layout of the help command.
- 8b873f3d5e4b Don't allow the user to add shortcuts whose name conflicts with commands.
- d63a1d032b89 Return a proper exit code.
- e55408103393 Using the =__g= namespace for everything.
- ebe139b64189 Allow relative paths.
- e5541ca8d93c Introduced CI.
- 21140b4d13d3 Print the usage function if no argument was given.
- 9e9bee407b7e Replaced the =realpath= function with one taken from [[https://github.com/travis-ci/gimme][travis-ci/gimme]].
- 0f7ff3a89637 Add copyright notice into the -v/--version flags.

** 0.2

- 3c070c44d7ef Tabulate output of the =list= command.
- 35c0df4330d4 The third argument of the =add= command is optional now.

** 0.1

- 50c8973995f1 Initial commit with basic functionality.
- 97122b6f5f29 Added the =--keys= option to the =list= command.
- 35176441ca34 Added a GNU Bash completion file.


t/test.sh
#!/bin/bash

set -e

source g.sh
cd /

# Setup the test file system.
mkdir -p a/b/c
mkdir -p a/d

g -v                                &>> /output.txt
g --version                         &>> /output.txt
g -h                                &>> /output.txt
g --help                            &>> /output.txt

# Let's add some shortcuts.
g add root
g add b a/b
g add c a/b/c

# Errors on the `add` command.
set +e
g add add                           &>> /output.txt
g add lala lala lala                &>> /output.txt
set -e

# Removing shortcuts.
g rm c
g rm d

# Errors on the `rm` command.
set +e
g rm                                &>> /output.txt
g rm lala lala                      &>> /output.txt
set -e

# The list command.
g list --keys | xargs -n1 | sort | xargs   &>> /output.txt
g list | awk '{ print $1 }' | sort | xargs &>> /output.txt

# Bare g command.
cd /
g b
basename $(pwd)                     &>> /output.txt
g root
basename $(pwd)                     &>> /output.txt
g root/a/b/c
basename $(pwd)                     &>> /output.txt

# Unknown shortcut.
set +e
g unknown                           &>> /output.txt
set -e

# Final diff
diff /expected.txt /output.txt


t/tests.sh
# Copyright (C) 2013-2016 Miquel Sabaté Solà <mikisabate@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

set -e

# From g.sh
__g_realpath() {
    [ -d "$1" ] && echo "$(cd "$1" && pwd)" || echo "$(cd "$(dirname "$1")" \
        && pwd)/$(basename "$1")"
}

# Cleanup previous tests and setup some special variables.
d="$(__g_realpath $(dirname $0))"
rm -f $d/output.txt
HOME="$d"

# Setup the test file system.
mkdir -p $d/a/b/c
mkdir -p $d/a/d

source $d/../g.sh

# First check that the basic commands are there.
g -v                                &>> $d/output.txt
g --version                         &>> $d/output.txt
g -h                                &>> $d/output.txt
g --help                            &>> $d/output.txt

# Let's add some shortcuts.
cd $d
g add root
g add b a/b
g add c a/b/c

# Errors on the `add` command.
set +e
g add add                           &>> $d/output.txt
g add lala lala lala                &>> $d/output.txt
set -e

# Removing shortcuts.
g rm c
g rm d

# Errors on the `rm` command.
set +e
g rm                                &>> $d/output.txt
g rm lala lala                      &>> $d/output.txt
set -e

# The list command.
g list --keys                       &>> $d/output.txt
g list | awk '{ print $1 }' | xargs &>> $d/output.txt

# Bare g command.
original=`pwd`
cd /
g b
basename $(pwd)                     &>> $d/output.txt
g root
basename $(pwd)                     &>> $d/output.txt
g root/a/b/c
basename $(pwd)                     &>> $d/output.txt
cd $original

# Unknown shortcut.
set +e
g unknown                           &>> $d/output.txt
set -e

# Teardown
diff $d/output.txt $d/expected.txt
rm $d/output.txt
rm -rf a
rm -rf $d/.gfile


run-tests.sh
# Copyright (C) 2013-2015 Miquel Sabaté Solà <mikisabate@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

set -e

if [ -z "$(pidof docker)" ]; then
    echo "Docker has to be running!"
    exit 1
fi
if [ ! -z "$(docker images g | grep latest)" ]; then
    id="$(docker images g | grep latest | awk '{ print $3 }')"
    docker rmi $id
fi

docker build -t g:latest .
docker run --rm -it g:latest


gcompletion.sh
# Copyright (C) 2013-2020 Miquel Sabaté Solà <mikisabate@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


# Do the completion work. Shamelessly taken from the git completion script.
__gcomp()
{
    local all c s=$'\n' IFS=' '$'\t'$'\n'
    local cur="${COMP_WORDS[COMP_CWORD]}"

    for c in $1; do
        case "$c$4" in
        --*=*) all="$all$c$4$s" ;;
        *.)    all="$all$c$4$s" ;;
        *)     all="$all$c$4 $s" ;;
        esac
    done
    IFS=$s
    COMPREPLY=($(compgen -P "$2" -W "$all" -- "$cur"))
    return
}

# Main function for the completion of the g command.
_g()
{
    local c=1 command
    opts=$(g list --keys)

    while [ $c -lt $COMP_CWORD ]; do
        command="${COMP_WORDS[c]}"
        c=$((++c))
    done

    if [ $c -eq $COMP_CWORD -a -z "$command" ]; then
        case "${COMP_WORDS[COMP_CWORD]}" in
        --*)   __gcomp "--version --help" ;;
        *)     __gcomp "${opts}" ;;
        esac
        return
    fi

    case "$command" in
    rm)         __gcomp "${opts}" ;;
    *) COMPREPLY=() ;;
    esac
}


complete -o default -o nospace -F _g g


g.sh
#!/bin/bash
# Copyright (C) 2013-2020 Miquel Sabaté Solà <mikisabate@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# Print the usage string.
__g_usage() {
    echo "usage: g [-h | --help] [-v | --version] <command> [<args>]"
}

# Print the help message.
__g_help() {
    __g_usage
    echo -e "\nThe g commands are:"
    tab=$'\t'
    column -t -s $'\t' <<HERE
   add${tab}Add a new shortcut.
   rm${tab}Remove a shorcut.
   list${tab}List all the available shortcuts.
HERE
}

# Fill the hash of shortcuts with the contents of the __g_file.
__g_get_shortcuts() {
    local word=""

    __g_shortcuts=()
    while read -r line; do
        if [ -n "$line" ]; then
            if [ -z "$word" ]; then
                word=$line
            else
                __g_shortcuts[$word]=$(eval echo "$line")
                word=""
            fi
        fi
    done < "$__g_file"
}

# Save the computed shortcuts into the __g_file.
__g_save_shortcuts() {
    # Erase the contents of the __g_file.
    :>"$__g_file"

    # Bash vs zsh
    if [ -n "$ZSH_VERSION" ]; then
        # shellcheck disable=SC2154
        keys="${(@i)__g_shortcuts}"
    else
        # shellcheck disable=SC2124
        keys="${!__g_shortcuts[@]}"
    fi

    # Finally write the hash into the __g_file.
    for i in $keys; do
        echo "$i" >> "$__g_file"
        echo "${__g_shortcuts[$i]}" >> "$__g_file"
    done
}

# Returns true if the given parameter is a keyword, false otherwise.
__g_is_keyword() {
    case "$1" in
    add | rm | list | -v | --version | -h | --help)
        return 0
        ;;
    *)
        return 1
    esac
}

# It returns a string with all the elements of the given array joined by slash
# characters.
__g_join_path() {
    local IFS="/"; echo "$*";
}

# Replacement for the non-standard `realpath` command. Implemented taken from
# https://github.com/travis-ci/gimme.
__g_realpath() {
    # shellcheck disable=SC2005
    [ -d "$1" ] && echo "$(cd "$1" && pwd)" || echo "$(cd "$(dirname "$1")" \
        && pwd)/$(basename "$1")"
}

# The main function for this script.
g() {
    version="0.3.9"
    cmd="$1"
    if [ -z "$cmd" ]; then
        __g_help
        return 1
    fi

    __g_file=$HOME/.gfile
    declare -A __g_shortcuts

    # Make sure that the __g_file actually exists.
    if [ -n "$GFILE" ]; then
        __g_file="$GFILE"
    fi
    touch "$__g_file"

    # Parse the command.
    case "$cmd" in
    -v | --version)
        cat <<HERE
g version $version
Copyright (C) 2013-2020 Miquel Sabaté Solà <mikisabate@gmail.com>
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
HERE
        ;;
    -h | --help)
        __g_help
        ;;
    add)
        if [ "$#" = "2" ]; then
            path=$(pwd)
        else
            if [ "$#" = "3" ]; then
                path="$3"
            else
                echo "usage: g add <name> [path]"
                return 1
            fi
        fi
        if __g_is_keyword "$2"; then
            echo "Cannot use '$2': keyword."
            return 1
        fi
        __g_get_shortcuts
        __g_shortcuts[$2]=$(__g_realpath "$path")
        __g_save_shortcuts
        ;;
    rm)
        if [ "$#" != "2" ]; then
            echo "usage: g rm <name>"
            return 1
        fi
        __g_get_shortcuts
        unset "__g_shortcuts[$2]"
        __g_save_shortcuts
        ;;
    list)
        __g_get_shortcuts
        if [[ "$#" = "2" && "$2" = "--keys" ]]; then
            str=""
            for i in "${!__g_shortcuts[@]}"; do
                str="$str $i"
            done
            echo "$str"
        else
            for i in "${!__g_shortcuts[@]}"; do
                echo -e "$i\t=> ${__g_shortcuts[$i]}"
            done | sort | column -t -s $'\t'
        fi
        ;;
    *)
        __g_get_shortcuts

        # Split the path and check whether the first element is a shortcut or
        # not.
        IFS='/' read -r -a path <<< "$cmd"
        init="${path[0]}"

        if [ -z "${__g_shortcuts[$init]}" ]; then
            echo -e "Unknown shortcut \`$init'.\n"
            __g_usage
            return 1
        else
            # Expand the shortcut and append the remaining parts of the path.
            path[0]="${__g_shortcuts[$init]}"
            cd "$(__g_join_path "${path[@]}")" || return 1
        fi
        ;;
    esac

    return 0
}
